wit/tls.wit

```diff
@@ -0,0 +1,309 @@
+        /// 
+        /// Returns `error(already-suspended)` is already suspended.
+        ///
+        /// Returns `error(closed)` if the connection has shut down either
+        /// successfully or abornmally.
+        ///
+        /// The suspension resource is a child resource of the client. Dropping
+        /// the `client` while it still has an active suspension resource may trap.
+        suspend: func() -> result<client-suspension, suspend-error>;
+
+        /// Resume the suspended client. Returns an error if the client is not
+        /// suspended.
+        /// 
+        /// If the client was suspended though `suspend` (i.e. it is not the
+        /// initial resumption), this method will trap if the suspension
+        /// resource hasn't been dropped yet.
+        ///
+        /// Returns an error if the client is not suspended.
+        resume: func() -> result;
+
+        /// Create a `pollable` which can be used to poll for
+        /// the client to be suspendable or closed.
+        ///
+        /// `subscribe` only has to be called once per client and can be (re)used
+        /// for the remainder of the client's lifetime.
+        subscribe: func() -> pollable;
+    }
+
+    /// The I/O streams that represent both sides of the transform.
+    ///
+    /// The application side interacts with the cleartext "private" streams.
+    /// The network side interacts with the encrypted "public" streams.
+    ///
+    /// A typical setup looks something like this:
+    /// 
+    /// ```text
+    /// :    TCP Socket                              TLS Client/Server
+    /// +-----------------+            +---------------------------------------------+
+    /// |                 |   splice   |                 decryption                  |   read
+    /// |  `input-stream` | ========>> | `public-output` =========>> `private-input` | ======>>   your
+    /// |                 |            |                                             |            app
+    /// |                 |            |                                             |            lives
+    /// | `output-stream` | <<======== | `public-input` <<========= `private-output` | <<======   here
+    /// |                 |   splice   |                 encryption                  |   write
+    /// +-----------------+            +---------------------------------------------+
+    /// ```
+    ///
+    /// The user of this interface is responsible for continually forwarding
+    /// data from the socket into the `public-output` stream and
+    /// data from the `public-input` into the socket.
+    ///
+    /// # Caution
+    /// Because the guest acts as both the producer and the consumer for these
+    /// streams, do not use the `blocking_*` methods as that will deadlock yourself.
+    record io-streams {
+        public-input: input-stream,
+        public-output: output-stream,
+        private-output: output-stream,
+        private-input: input-stream,
+    }
```

Comment:
That's what I had initially. It would definitely be a simpler interface. However, WASI guests are not able to create arbitrary input/output-streams themselves. Consuming the input+output stream would effectively mean that the TLS interface can **_only_** be used directly on top of TCP sockets. In theory this sounds fine for an initial version, but is likely a non-starter for the .NET folks who need raw access to the TLS data, because their [primary TLS abstraction](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslstream?view=net-8.0) has been designed that way.

Reply:
Could you elaborate to help me understand the incompatibility with .NET?

I'm not too familiar with that ecosystem, but looking at the docs, the `SslStream` takes an abstract `Stream` as a parameter https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.-ctor?view=net-8.0

Such an abstract `Stream`, it seems, would need to merge both `input-stream` and `output-stream` into a single structure. Assuming .NET standard library supports WASI, following the linked examples:

This would return a client structure, which would embed the input and output stream, received from `wasi:sockets`
```c#
TcpClient client = listener.AcceptTcpClient();
```

`client.GetStream()` would return a structure, which embeds both input and output stream
```c#
 SslStream sslStream = new SslStream(
                client.GetStream(), false);
```

Also, not entirely sure why we'd need to create arbitrary streams here.